src/main/scala/rps/persistence/GameRepository.scala

```diff
@@ -0,0 +1,50 @@
+package rps
+
+import io.buildo.enumero.CaseEnumSerialization
+import java.util.UUID
+import java.sql.Timestamp
+import java.time.Instant
+import slick.jdbc.JdbcBackend.DatabaseDef
+import slick.jdbc.PostgresProfile.api._
+import scala.concurrent.{ExecutionContext}
+import scala.concurrent.Future
+import db.Tables.{Match, Row}
+import models._
+
+trait GameRepository {
+  def save(play: Play): Future[UUID]
+
+  def readLastMatch(): Future[Option[Play]]
+}
+
+object GameRepository {
+  def create(db: DatabaseDef)(implicit ec: ExecutionContext) =
+    new GameRepository {
+      val moveSerializer = CaseEnumSerialization[Move]
+      val resultSerializer = CaseEnumSerialization[GameResult]
+
+      def save(play: Play): Future[UUID] = {
+        val rowId = UUID.randomUUID
+        val row = Row(
+          id = rowId,
+          computerMove = moveSerializer.caseToString(play.computerMove),
+          userMove = moveSerializer.caseToString(play.userMove),
+          result = resultSerializer.caseToString(play.result),
+          occurredAt = Timestamp.from(Instant.now)
+        )
+        db.run(Match += row).map(_ => rowId)
+      }
+
+      def readLastMatch(): Future[Option[Play]] =
+        db.run(Match.sortBy(_.occurredAt.desc).take(1).result.headOption)
+          .map(_.flatMap(convertPlayRow)) // TODO approfondire cosa fa questa riga
```

Comment:
qui è un po' contorto perché ci sono più livelli di nesting

`db.run(...)` ha tipo `Future[Option[Row]]`

con la `.map` dici che vuoi lavorare sul risultato della future, ma con un'operazione sincrona, non asincrona, quindi non ti serve una `flatMap` ma basta una `map`

alla map devi passare una funzione di tipo `Option[Row] => Option[Play]`

se avessimo una funzione `convertPlayRow` di tipo `Row => Play`, potremmo di nuovo usare una map, quindi scrivere `.map(_.map(convertPlayRow))`

però la vera `convertPlayRow` ha invece tipo `Row => Option[Play]` perché la conversione può fallire (se gli enum salvati a DB non corrispondono a quelli del codice –  caso che peraltro non dovrebbe capitare perché abbiamo definito le colonne come enum in Postgres), quindi usiamo `flatMap`

in tutto ciò non so se hai presente la sintassi `_`, non mi ricordo se l'abbiamo vista altrove: è un modo per scrivere funzioni in modo compatto, per cui `_.qualcosa(...)` è come scrivere `x => x.qualcosa(...)` cioè ti dà un modo compatto di scrivere una funzione quando questa usa il suo argomento una sola volta